hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/ServerRemoteProcedure.java

```diff
@@ -137,6 +140,10 @@ synchronized void remoteOperationDone(MasterProcedureEnv env, Throwable error) {
         getProcId());
       return;
     }
+    //below persistence is added so that if report goes to last active master, it throws exception
+    state = MasterProcedureProtos.ServerRemoteProcedureState.SERVER_REMOTE_PROCEDURE_REPORT_SUCCEED;
+    env.getMasterServices().getMasterProcedureExecutor().getStore().update(this);
+
     complete(env, error);
```

Comment:
Here, we should not call the complete directly again. We should wake up the procedure and process the result in the execute method.

Reply:
If I have to call the `complete` method in `execute` methods, I have to store the Error as `complete` methods are using that to take the decision. 

Another solution can be I play with the state. According to the error in `remoteOperationDone` method, I can change the state and then take all the decisions in `complete` method based on the state. But I don't think this would be possible without introducing some state specific to some parent procedure like SnapshotVerifyProcedure.

Reply:
Persisting the error is a must...
Think of this scenario, after you persist the state, master crashes before you call complete, and after restart, you enter the execute method, if you do not have the error object then how do you plan to call the comlete method...

Reply:
Any updates here?

Reply:
I was also thinking of the scenarios, where while releasing the new bits there might be two kinds of HMaster one with new bits and one with old bits. What will happen if Active Master with new bits crashes/stops and the Master with old bits becomes Active Master and vice-versa? 

Do we have some guidelines while releasing that I need to keep in mind while thinking of such scenarios?

Reply:
I tried to write another approach as well. You can take a look. https://github.com/apache/hbase/compare/master...Umeshkumar9414:hbase:HBASE_28420_2?expand=1

Reply:
The above link doesn't work @Umeshkumar9414

Reply:
I do not think you need to consider a master with old code become active, as we are trying to fix a bug in the old code...

And I still do not get your point on why you must call complete in remoteOperationDone method, as I explained above, if master crashes after we persist the state and then restarts, we need to call complete in the execute method, so why not just follow the same pattern since you must have the code in execute method?
And there is another advantage to prevent potential data race if you call complete method in execute method, as the execute method will be called by PEWorker, where we have a bunch of fencing ways to make sure that there is no race.

Reply:
I wanted to call complete in remoteOperationDone so that we didn't encounter the scenario of a master with old code becoming active. Anyway as we don't need to consider this, I have updated the PR with new changes.